### PR TITLE
[SPARK-22378][SQL] Eliminate redundant null check in generated code for extracting an element from complex types

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

This PR eliminates redundant null check in generated code for extracting an element from complex types `GetArrayItem`, `GetMapValue`, and `GetArrayStructFields`. Since these code generation does not take care of `nullable` in `DataType` such as `ArrayType`, the generated code always has `isNullAt(index)`.  
This PR avoids to generate `isNullAt(index)` if `nullable` is false in `DataType`.

Example
```
    val nonNullArray = Literal.create(Seq(1), ArrayType(IntegerType, false))
    checkEvaluation(GetArrayItem(nonNullArray, Literal(0)), 1)
```

Before this PR
```
/* 034 */   public java.lang.Object apply(java.lang.Object _i) {
/* 035 */     InternalRow i = (InternalRow) _i;
/* 036 */
/* 037 */
/* 038 */
/* 039 */     boolean isNull = true;
/* 040 */     int value = -1;
/* 041 */
/* 042 */
/* 043 */
/* 044 */     isNull = false; // resultCode could change nullability.
/* 045 */
/* 046 */     final int index = (int) 0;
/* 047 */     if (index >= ((ArrayData) references[0]).numElements() || index < 0 || ((ArrayData) references[0]).isNullAt(index)) {
/* 048 */       isNull = true;
/* 049 */     } else {
/* 050 */       value = ((ArrayData) references[0]).getInt(index);
/* 051 */     }
/* 052 */     isNull_0 = isNull;
/* 053 */     value_0 = value;
/* 054 */
/* 055 */     // copy all the results into MutableRow
/* 056 */
/* 057 */     if (!isNull_0) {
/* 058 */       mutableRow.setInt(0, value_0);
/* 059 */     } else {
/* 060 */       mutableRow.setNullAt(0);
/* 061 */     }
/* 062 */
/* 063 */     return mutableRow;
/* 064 */   }
```

After this PR (Line 47 is changed)
```
/* 034 */   public java.lang.Object apply(java.lang.Object _i) {
/* 035 */     InternalRow i = (InternalRow) _i;
/* 036 */
/* 037 */
/* 038 */
/* 039 */     boolean isNull = true;
/* 040 */     int value = -1;
/* 041 */
/* 042 */
/* 043 */
/* 044 */     isNull = false; // resultCode could change nullability.
/* 045 */
/* 046 */     final int index = (int) 0;
/* 047 */     if (index >= ((ArrayData) references[0]).numElements() || index < 0) {
/* 048 */       isNull = true;
/* 049 */     } else {
/* 050 */       value = ((ArrayData) references[0]).getInt(index);
/* 051 */     }
/* 052 */     isNull_0 = isNull;
/* 053 */     value_0 = value;
/* 054 */
/* 055 */     // copy all the results into MutableRow
/* 056 */
/* 057 */     if (!isNull_0) {
/* 058 */       mutableRow.setInt(0, value_0);
/* 059 */     } else {
/* 060 */       mutableRow.setNullAt(0);
/* 061 */     }
/* 062 */
/* 063 */     return mutableRow;
/* 064 */   }
```

## How was this patch tested?

Added test cases into `ComplexTypeSuite`

### CODE REVIEW

**Comment by kiszk** [comment]
@gatorsmile could you please review this?

**Comment by kiszk** [comment]
ping @gatorsmile

**Comment by kiszk** [comment]
ping @gatorsmile @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 1]
```diff
@@ -186,6 +186,7 @@ case class GetArrayStructFields(
       val values = ctx.freshName("values")
       val j = ctx.freshName("j")
       val row = ctx.freshName("row")
+      val nullCheckElement = if (!field.nullable) "false" else s"$row.isNullAt($ordinal)"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: how about
```
val nullSafeEval = if (field.nullable) {
  s"""
    if ($row.isNullAt($ordinal)) {
      $values[$j] = null;
    } else
  """
} else ""
...
...

$nullSafeEval {
  $values[$j] = ${ctx.getValue(row, field.dataType, ordinal.toString)};
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 2]
```diff
@@ -242,9 +243,11 @@ case class GetArrayItem(child: Expression, ordinal: Expression)
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     nullSafeCodeGen(ctx, ev, (eval1, eval2) => {
       val index = ctx.freshName("index")
+      val nullCheck = if (!child.dataType.asInstanceOf[ArrayType].containsNull) ""
+        else s" || $eval1.isNullAt($index)"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: nit:
```
if {

} else {

}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 3]
```diff
@@ -309,6 +312,8 @@ case class GetMapValue(child: Expression, key: Expression)
     val found = ctx.freshName("found")
     val key = ctx.freshName("key")
     val values = ctx.freshName("values")
+    val nullCheck = if (!child.dataType.asInstanceOf[MapType].valueContainsNull) ""
+      else s" || $values.isNullAt($index)"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: ditto

**Comment by cloud-fan** [comment]
LGTM

**Comment by kiszk** [comment]
Jenkins, retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 4]
```diff
@@ -186,6 +186,14 @@ case class GetArrayStructFields(
       val values = ctx.freshName("values")
       val j = ctx.freshName("j")
       val row = ctx.freshName("row")
+      val nullSafeEval = if (field.nullable) {
+        s"""
+         if ($row.isNullAt($ordinal)) {
+           $values[$j] = null;
+         } else
+        """
+      } else ""
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: Nit:
```Scala
} else {
  ""
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 5]
```diff
@@ -309,6 +318,9 @@ case class GetMapValue(child: Expression, key: Expression)
     val found = ctx.freshName("found")
     val key = ctx.freshName("key")
     val values = ctx.freshName("values")
+    val nullCheck = if (child.dataType.asInstanceOf[MapType].valueContainsNull) {
+      s" || $values.isNullAt($index)"
+    } else ""
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: The same here

**Comment by gatorsmile** [comment]
LGTM

BTW, now, more and more optimization are added based on the NULLable property in schema. Normally, the database enforces null checking during inserting/updating the data, but we do not enforce it because the data is not managed by us. If the users provide wrong nullable info in the schema. This might be harder for us to debug these runtime issues.

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Comment by kiszk** [comment]
I see. It makes sense for debugability. On the other hand, the NULLable property in schema is very important for optimizations to achieve better performance.

One of ideas is to add conf for debug mode to enforce insertion of null check at reading data from an iterator or column vector.

**Comment by gatorsmile** [comment]
Maybe just changing the schema with always-on nullability

**Pull request closed without merging**